fe/fe-core/src/main/java/com/starrocks/qe/scheduler/assignment/LocalFragmentAssignmentStrategy.java

```diff
@@ -0,0 +1,305 @@
+import com.starrocks.common.util.ListUtil;
+import com.starrocks.planner.PlanFragment;
+import com.starrocks.planner.ScanNode;
+import com.starrocks.qe.BackendSelector;
+import com.starrocks.qe.ColocatedBackendSelector;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.qe.scheduler.WorkerProvider;
+import com.starrocks.qe.scheduler.dag.ExecutionFragment;
+import com.starrocks.qe.scheduler.dag.FragmentInstance;
+import com.starrocks.system.ComputeNode;
+import com.starrocks.thrift.TScanRangeParams;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The assignment strategy for fragments whose left most node is a scan node.
+ * <p> It firstly assigns scan ranges to workers, and then dispatches scan ranges, assigned to each worker, to fragment instances.
+ */
+public class LocalFragmentAssignmentStrategy implements FragmentAssignmentStrategy {
+    private static final Logger LOG = LogManager.getLogger(LocalFragmentAssignmentStrategy.class);
+
+    private final ConnectContext connectContext;
+    private final WorkerProvider workerProvider;
+    private final boolean usePipeline;
+    private final boolean isLoadType;
+
+    private final Set<Integer> replicatedScanIds = Sets.newHashSet();
+
+    public LocalFragmentAssignmentStrategy(ConnectContext connectContext, WorkerProvider workerProvider, boolean usePipeline,
+                                           boolean isLoadType) {
+        this.connectContext = connectContext;
+        this.workerProvider = workerProvider;
+        this.usePipeline = usePipeline;
+        this.isLoadType = isLoadType;
+    }
+
+    @Override
+    public void assignFragmentToWorker(ExecutionFragment execFragment) throws UserException {
+        for (ScanNode scanNode : execFragment.getScanNodes()) {
+            assignScanRangesToWorker(execFragment, scanNode);
+        }
+
+        assignScanRangesToFragmentInstancePerWorker(execFragment);
+
+        // The fragment which only contains scan nodes without scan ranges,
+        // such as SchemaScanNode, is assigned to an arbitrary worker.
+        if (execFragment.getInstances().isEmpty()) {
+            long workerId = workerProvider.selectNextWorker();
+            ComputeNode worker = workerProvider.getWorkerById(workerId);
+            FragmentInstance instance = new FragmentInstance(worker, execFragment);
+            execFragment.addInstance(instance);
+        }
+    }
+
+    private void assignScanRangesToWorker(ExecutionFragment execFragment, ScanNode scanNode) throws UserException {
+        BackendSelector backendSelector = BackendSelectorFactory.create(
+                scanNode, isLoadType, execFragment, workerProvider, connectContext, replicatedScanIds);
+        if (backendSelector == null) {
+            return;
```

Comment:
No, it isn't an error. some ScanNodes doesn't have scan ranges, such as SchemaScanNode will be assigned to an arbitrary worker.
I would also add some comments here.

```java
    @Override
    public void assignFragmentToWorker(ExecutionFragment execFragment) throws UserException {
        for (ScanNode scanNode : execFragment.getScanNodes()) {
            assignScanRangesToWorker(execFragment, scanNode);
        }

        assignScanRangesToFragmentInstancePerWorker(execFragment);

        // ===== here =====
        // The fragment which only contains scan nodes without scan ranges,
        // such as SchemaScanNode, is assigned to an arbitrary worker.
        if (execFragment.getInstances().isEmpty()) {
            long workerId = workerProvider.selectNextWorker();
            ComputeNode worker = workerProvider.getWorkerById(workerId);
            FragmentInstance instance = new FragmentInstance(worker, execFragment);
            execFragment.addInstance(instance);
        }
    }
```

Reply:
You can provide a trivial backendSelector for this cases to do nothing instead of null. by this way, backendSelector would be a total concept instead of partial concept.

Reply:
Great! It is consistent and maintained more easily. Add `NoopBackendSelector`. Is there any suggestion for a better class name of `NoopBackendSelector`?